File: Euclid/Network/GameNetworkHandler.cs
using DotNetty.Common.Utilities;
using DotNetty.Transport.Channels;
using Euclid.Messages;
using Euclid.Messages.Outgoing;
using Euclid.Network.Session;
using Euclid.Network.Streams;
using log4net;
using System;

namespace Euclid.Network
{
    internal class GameNetworkHandler : ChannelHandlerAdapter
    {
        #region Fields

        public static AttributeKey<ConnectionSession> CONNECTION_KEY = AttributeKey<ConnectionSession>.NewInstance("CONNECTION_KEY");
        public static readonly ILog log = LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);

        #endregion

        #region Public methods

        /// <summary>
        /// Handle client connections.
        /// </summary>
        /// <param name="ctx">the channel context</param>
        public override void ChannelActive(IChannelHandlerContext ctx)
        {
            base.ChannelActive(ctx);

            var connection = new ConnectionSession(ctx.Channel);

            if (connection != null)
            {
                connection.Player.Log.Debug($"Client connected to server: {connection.IpAddress}");
                ctx.Channel.GetAttribute(CONNECTION_KEY).SetIfAbsent(connection);

                ctx.Channel.WriteAndFlushAsync(new HELLO());
            }
        }

        /// <summary>
        /// Handle client disconnects.
        /// </summary>
        /// <param name="ctx">the channel context</param>
        public override void ChannelInactive(IChannelHandlerContext ctx)
        {
            base.ChannelInactive(ctx);
            ConnectionSession connection = ctx.Channel.GetAttribute(CONNECTION_KEY).Get();

            if (connection == null)
                return;

            connection.OnDisconnect();

            connection.Player.Log.Debug($"Client disconnected from server: {connection.IpAddress}");
        }

        /// <summary>
        /// Handle incoming channel messages from the decoder
        /// </summary>
        /// <param name="ctx">the channel context</param>
        /// <param name="msg">the incoming message</param>
        public override void ChannelRead(IChannelHandlerContext ctx, object msg)
        {
            ConnectionSession connectionSession = ctx.Channel.GetAttribute(CONNECTION_KEY).Get();

            if (connectionSession == null)
                return;

            if (msg is Request)
            {
                Request request = (Request)msg;
                MessageHandler.Instance.HandleMesage(connectionSession.Player, request);
            }

            base.ChannelRead(ctx, msg);
        }

        /// <summary>
        /// Handle channel read complete.
        /// </summary>
        /// <param name="context">the channel context</param>
        public override void ChannelReadComplete(IChannelHandlerContext context) => context.Flush();

        /// <summary>
        /// Handle exceptions thrown by the network api.
        /// </summary>
        /// <param name="context">the channel context</param>
        /// <param name="exception">the exception</param>
        public override void ExceptionCaught(IChannelHandlerContext context, Exception exception) =>
            log.Error(exception.ToString());

        #endregion
    }
}

File: euclid/src/main/java/euclid/network/GameNetworkHandler.java
package euclid.network;

import euclid.messages.IMessageComposer;
import euclid.messages.MessageHandler;
import euclid.network.session.ConnectionSession;
import euclid.network.streams.Request;
import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.SimpleChannelInboundHandler;
import io.netty.util.AttributeKey;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public class GameNetworkHandler extends SimpleChannelInboundHandler<Object> {

    public static final AttributeKey<ConnectionSession> CONNECTION_KEY =
            AttributeKey.valueOf("CONNECTION_KEY");

    private static final Logger log = LogManager.getLogger(GameNetworkHandler.class);

    @Override
    public void channelActive(ChannelHandlerContext ctx) {
        ConnectionSession connection = new ConnectionSession(ctx.channel());

        connection.getPlayer().getLog().debug("Client connected to server: {}", connection.getIpAddress());
        ctx.channel().attr(CONNECTION_KEY).setIfAbsent(connection);

        ctx.channel().writeAndFlush(new euclid.messages.outgoing.HELLO());
    }

    @Override
    public void channelInactive(ChannelHandlerContext ctx) {
        ConnectionSession connection = ctx.channel().attr(CONNECTION_KEY).get();

        if (connection == null) {
            return;
        }

        connection.onDisconnect();
        connection.getPlayer().getLog().debug("Client disconnected from server: {}", connection.getIpAddress());
    }

    @Override
    protected void channelRead0(ChannelHandlerContext ctx, Object msg) {
        ConnectionSession connectionSession = ctx.channel().attr(CONNECTION_KEY).get();

        if (connectionSession == null) {
            return;
        }

        if (msg instanceof Request request) {
            MessageHandler.getInstance().handleMessage(connectionSession.getPlayer(), request);
        }
    }

    @Override
    public void channelReadComplete(ChannelHandlerContext ctx) {
        ctx.flush();
    }

    @Override
    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
        log.error(cause.toString());
    }
}
